File: solutions/python/stack/largest-rectangle-in-histogram.py
# LeetCode 84 — Largest Rectangle in Histogram (Hard)
# Category: Stack · Approach: Monotonic stack
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/largest-rectangle-in-histogram/

def largest_rectangle_area(heights: list[int]) -> int:
    stack: list[int] = []
    max_area = 0
    n = len(heights)
    for i in range(n + 1):
        h = 0 if i == n else heights[i]
        while stack and heights[stack[-1]] >= h:
            top = stack.pop()
            height = heights[top]
            width = i - stack[-1] - 1 if stack else i
            max_area = max(max_area, height * width)
        stack.append(i)
    return max_area

File: solutions/typescript/stack/largest-rectangle-in-histogram.ts
// LeetCode 84 — Largest Rectangle in Histogram (Hard)
// Category: Stack · Approach: Monotonic stack
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/largest-rectangle-in-histogram/

function largestRectangleArea(heights: number[]): number {
  const stack: number[] = [];
  let maxArea = 0;
  const n = heights.length;
  for (let i = 0; i <= n; i++) {
    const h = i === n ? 0 : heights[i];
    while (stack.length && heights[stack[stack.length - 1]] >= h) {
      const top = stack.pop()!;
      const height = heights[top];
      const width = stack.length ? i - stack[stack.length - 1] - 1 : i;
      maxArea = Math.max(maxArea, height * width);
    }
    stack.push(i);
  }
  return maxArea;
}
